### PR TITLE
Make code gen generate fragments as pointers

### PR DESCRIPTION
**Pull request by tian000**
https://github.com/Yamashou/gqlgenc/issues/60

### CODE REVIEW

File: example/fragment/gen/client.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		Options []*struct {
+			ID   string "json:\"id\" graphql:\"id\""
+			Text string "json:\"text\" graphql:\"text\""
+		} "json:\"options\" graphql:\"options\""
+	} "json:\"question\" graphql:\"question\""
+}
+type ConversationByIDQuery struct {
+	Conversation *struct {
+		Messages []*struct {
+			Answer *struct {
```
Comment (tian000) [inline example/fragment/gen/client.go:42]: @adam-zethraeus is this how you want the generated code to look like?

File: clientgenv2/source_generator.go [hunk 2]
```diff
@@ -148,7 +148,7 @@ func (r *SourceGenerator) NewResponseField(selection ast.Selection) *ResponseFie
 		case fieldsResponseFields.IsFragment():
 			// 子フィールドがFragmentの場合はこのFragmentがフィールドの型になる
 			// if a child field is fragment, this field type became fragment.
-			baseType = fieldsResponseFields[0].Type
+			baseType = types.NewPointer(fieldsResponseFields[0].Type)
```
Comment (adam-zethraeus) [inline clientgenv2/source_generator.go:151]: will this skip adding a second indirect if the fragment is directly on a `non-!` field?

File: example/fragment/gen/client.go (hunk 1 above)
Comment (adam-zethraeus) [reply example/fragment/gen/client.go:42]: Yes! i think that's a substantive improvement. At least now we can test the payload :).

**Review by adam-zethraeus — COMMENTED** [review]
for reference, apollo-tools's client codegen has a few really nice properties that gqlgenc doesn't:
* it doesn't distinguish between models and payload. i.e:
* it doesn't generate models which aren't used in queries
* if generates only the parts of models which *are* used in queries.
* it flattens inline fragments properly into their hosting type

in general it seems like gqlgenc is just using gqlgen's model gen — and that model gen just mirrors the *schema*. That's great for a server, but it doesn't suit the client.

AFAICT, there's no way to go from a gqlgenc  Payload to a Model in any but the most simple cases!

This is made even more frustrating by the Payload being comprised of anonymous structs. I'd like to use those structs!

Happy to approve this. LMK. The above is just my general thoughts. (I don't have time to dig in on improving this codegen right now, but would more generally be interested in working on it w/ you too!)

**Comment by tian000** [comment]
I think its better that I close this PR. Since it probably doesn't cover other cases, e.g. where the types are already pointers.
"Regarding some comments about this library:

1. it doesn't distinguish between models and payload. i.e:
I think we should delete the generated models since they are unused.

2. it doesn't generate models which aren't used in queries
Solved by deleting the models

3. it generates only the parts of models which are used in queries.
gqlgenc already does this, its just that the structs are anonymous. We could make it non-anonymous, i think this change is do-able.

4. it flattens inline fragments properly into their hosting type
I don't know what this means exactly. Some of this awkwardness might just be due to the fact that is not a native way to represent sum types in golang.

**Comment by adam-zethraeus** [comment]
> I think its better that I close this PR. Since it probably doesn't cover other cases, e.g. where the types are already pointers.
> "Regarding some comments about this library:
> 
> 1. it doesn't distinguish between models and payload. i.e:
>    I think we should delete the generated models since they are unused.

makes sense to me.

> 2. it doesn't generate models which aren't used in queries
>    Solved by deleting the models
> 3. it generates only the parts of models which are used in queries.
>    gqlgenc already does this, its just that the structs are anonymous. We could make it non-anonymous, i think this change is do-able.

making them non-anonymous would be a big win, yes

> 4. it flattens inline fragments properly into their hosting type
>    I don't know what this means exactly. Some of this awkwardness might just be due to the fact that is not a native way to represent sum types in golang.

```
        answer {
        thing
        ... on ChatSingleSelectAnswer {
            selectionId
        }
 ```
 
 currently this is represented as:
 ```
type Answer struct {
    Thing string
    ChatSingleSelectAnswer struct {
        SelectionID string
    }
}
 ```
 
 the representation that would correspond to the schema would be:
  currently this is represented as:
 ```
type ChatSingleSelectAnswer struct {
        Thing string
        SelectionID string
}
 ```

**Comment by adam-zethraeus** [comment]
thanks for digging into this!

**Pull request closed without merging**